pkg/middlewares/tracing/wrapper.go

```diff
@@ -21,48 +21,99 @@ func Wrap(ctx context.Context, constructor alice.Constructor) alice.Constructor
 		if constructor == nil {
 			return nil, nil
 		}
-		handler, err := constructor(next)
+		// we need to start tracing when middleware (provided by `constructor`) starts
+		// and finish when middleware ends, but before calling `next`
+		// so we need to wrap `next` into TraceFinisher
+		// and call TraceFinisher when middleware processed
+		tracingFinisher := NewTraceFinisher(next)
+		handler, err := constructor(tracingFinisher)
 		if err != nil {
 			return nil, err
 		}
 
 		if tracableHandler, ok := handler.(Tracable); ok {
 			name, spanKind := tracableHandler.GetTracingInformation()
 			log.FromContext(ctx).WithField(log.MiddlewareName, name).Debug("Adding tracing to middleware")
-			return NewWrapper(handler, name, spanKind), nil
+			return NewTraceStarter(handler, name, spanKind, tracingFinisher), nil
 		}
 		return handler, nil
 	}
 }
 
-// NewWrapper returns a http.Handler struct
-func NewWrapper(next http.Handler, name string, spanKind ext.SpanKindEnum) http.Handler {
-	return &Wrapper{
-		next:     next,
-		name:     name,
-		spanKind: spanKind,
+// NewTraceStarter returns a *TracingStarted struct
+func NewTraceStarter(middleware http.Handler, name string, spanKind ext.SpanKindEnum, finisher *TraceFinisher) *TraceStarter {
```

Comment:
Neither this constructor, nor any of the new types in this file, need to be exported, do they?